string/src/io/github/iltotore/iron/string/constraint.scala

```diff
@@ -12,40 +13,53 @@ object constraint {
    */
   trait LowerCase
 
-  inline given Constraint.RuntimeOnly[String, LowerCase] with {
-    override inline def assert(value: String): Boolean = value equals value.toLowerCase
+  inline given Constraint[String, LowerCase] with {
+    override inline def assert(value: String): Boolean = compileTime.checkLowerCase(value)
 
-    override inline def getMessage(value: String): String = s"$value should be lower cased"
+    override inline def getMessage(value: String): String = value + " should be lower cased"
   }
 
   /**
    * Constraint: checks if the input value is upper case.
    */
   trait UpperCase
 
-  inline given Constraint.RuntimeOnly[String, UpperCase] with {
-    override inline def assert(value: String): Boolean = value equals value.toUpperCase
+  inline given Constraint[String, UpperCase] with {
+    override inline def assert(value: String): Boolean = compileTime.checkUpperCase(value)
 
-    override inline def getMessage(value: String): String = s"$value should be upper cased"
+    override inline def getMessage(value: String): String = value + " should be upper cased"
   }
 
   /**
    * Constraint: checks if the input value matches V.
    * @tparam V the regex to match with.
    */
-  trait Match[V]
+  trait Match[V <: String & Singleton]
 
   type Alphanumeric = Match["^[a-zA-Z0-9]+"] DescribedAs "Value should be alphanumeric"
 
+  inline given MatchConstraint[Alphanumeric] with {}
+
   type URLLike =
     Match["^(?:http(s)?:\\/\\/)?[\\w.-]+(?:\\.[\\w\\.-]+)+[\\w\\-\\._~:/?#\\[\\]@!\\$&'\\(\\)\\*\\+,;=.]+$"] DescribedAs "Value should be an URL"
 
+  inline given MatchConstraint[URLLike] with {}
+
   type UUIDLike =
     Match["^([0-9a-fA-F]{8}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{12})"] DescribedAs "Value should be an UUID"
 
-  inline given [V <: String]: Constraint.RuntimeOnly[String, Match[V]] with {
+  inline given MatchConstraint[UUIDLike] with {}
+
+  trait MatchConstraint[M <: Match[_] | DescribedAs[Match[_], _]] extends Constraint[String, M] {
```

Comment:
You're `Constraint` instance definition seems to be unusual. Couldn't you use a definition like
```scala
class MatchConstraint[V <: String] extends Constraint[String, Match[V]] {

  override inline def assert(value: String): Boolean = compileTime.checkMatch(value, constValue[V])

  override inline def getMessage(value: String): String = value + " should match " + constValue[V]
} 
```

Reply:
The issue is that in MatchConstraint type parameters I want `Match`, because I'd like to pass `Match` to it.

It would be great If I could specify it like this:
```
class MatchConstraint[Match[V], V <: String] extends Constraint[String, Match[V]] {

  override inline def assert(value: String): Boolean = compileTime.checkMatch(value, constValue[V])

  override inline def getMessage(value: String): String = value + " should match " + constValue[V]
} 

// V is autoinferred
inline given MatchConstraint[UUIDLike] with {}
```

but unfortunatelly I cannot.